tests/rptest/services/kcat_consumer.py

```diff
@@ -0,0 +1,274 @@
+        chan = client.get_transport().open_session(timeout=timeout_sec)
+
+        chan.settimeout(timeout_sec)
+        chan.exec_command(cmd)
+        chan.set_combine_stderr(False)
+
+        stdin = chan.makefile('wb', -1)  # set bufsize to -1
+        stdout = chan.makefile('r', -1)
+        stderr = chan.makefile_stderr('r', -1)
+
+        return KcatConsumer.SSHCapturedPipes(
+            stdin, stdout, stderr, cmd,
+            self._redpanda.logger if allow_fail else None)
+
+        def output_generator():
+            for line in iter(stdout.readline, ''):
+                yield line
+
+        return SSHOutputIter(output_generator, stdout)
+
+    def _worker(self, worker_idx, node):
+        self._stopping.clear()
+        try:
+            cmd = "echo $$ ; " + " ".join(self._cmd)
+            with self.ssh_capture_with_pipes(node, cmd) as p:
+                stop_stderr_reader = threading.Event()
+                stderr_reader = threading.Thread(
+                    name=self.service_id + "-errrdr-" + str(worker_idx),
+                    target=self._stderr_reader,
+                    args=(
+                        worker_idx,
+                        node,
+                        #stop_stderr_reader,
+                        p.stderr))
+                stderr_reader.start()
+                try:
+                    self._stdout_reader(p.stdout)
+                finally:
+                    stop_stderr_reader.set()
+                    stderr_reader.join()
+
+        except:
+            if self._stopping.is_set():
+                # Expect a non-zero exit code when killing during teardown
+                pass
+            else:
+                raise
+        finally:
+            self.done = True
+
+    def _stdout_reader(self, stdout: paramiko.channel.ChannelFile) -> None:
+        #self._redpanda.logger.debug(f"start ({stdout})")
+        for line in iter(stdout.readline, ''):
+            if self._pid is None:
+                self._pid = line.strip()
+                continue
+
+            if self._stopping.is_set():
+                break
+
+            try:
+                j = json.loads(line)
+                partition = int(j["partition"])
+                self._end_of_topic[partition] = None
+                self._consumed_count.setdefault(partition, 0)
+                self._consumed_count[partition] += 1
+                self._on_message(self, j)
+            except:
+                self._redpanda.logger.error(
+                    f"{self._caption}Exception while processing kcat output line: {line.strip()}"
+                )
+                raise
+
+    def _stderr_reader(
+        self,
+        node_idx,
+        node,
+        #stop: threading.Event,
+        stderr: paramiko.channel.ChannelStderrFile
+    ) -> None:
+        #self._redpanda.logger.debug(f"start ({node_idx}, {node}, {stderr})")
+        for line in iter(stderr.readline, ''):
+            if self._stopping.is_set():
```

Comment:
This is probably because my python skills are lacking, but my question is: if there is no data coming through `stderr`, will line 222 just block and then you don't get alerted about self._stopping?

Reply:
That's not python skills, that's about `paramiko` library. `_stopping` is set only from `stop_node()`, followed there by killing the process. Once the process is killed, all waiting on its channel is unblocked, and we will see that the event is set.